xc/common/utils/prjuray_form_channels.py

```diff
@@ -0,0 +1,481 @@
+    """
+
+    R = 0.0
+    C = 0.0
+    Tdel = 0.0
+    penalty_cost = 0.0
+
+    # Timings
+    if pip_timing is not None:
+
+        if pip_timing.delays is not None:
+            # Use the largest intristic delay for now.
+            # This conservative on slack timing, but not on hold timing.
+            #
+            # nanosecond -> seconds
+            Tdel = pip_timing.delays[PvtCorner.SLOW].max / 1e9
+
+        if pip_timing.internal_capacitance is not None:
+            # microFarads -> Farads
+            C = pip_timing.internal_capacitance / 1e6
+
+        if pip_timing.drive_resistance is not None:
+            # milliOhms -> Ohms
+            R = pip_timing.drive_resistance / 1e3
+
+    return R, C, Tdel, penalty_cost
+
+
+def get_site_pin_timing(site_pin):
+    """
+    Returns internal R, C and Tdel for the given site pin
+    """
+
+    R = 0.0
+    C = 0.0
+
+    # Use the largest intristic delay for now.
+    # This conservative on slack timing, but not on hold timing.
+
+    # nanosecond -> seconds
+    Tdel = site_pin.timing.delays[PvtCorner.SLOW].max / 1e9
+
+    if isinstance(site_pin.timing, prjuray.tile.OutPinTiming):
+        # milliOhms -> Ohms
+        R = site_pin.timing.drive_resistance / 1e3
+    elif isinstance(site_pin.timing, prjuray.tile.InPinTiming):
+        # microFarads -> Farads
+        C = site_pin.timing.capacitance / 1e6
+    else:
+        assert False, site_pin
+
+    return R, C, Tdel
+
+
+def check_pip_for_direct(pip_name):
+    """
+    Returns true when the given pip should be modeled as a direct connection
+    """
+    return False
+
+
+def connect_hardpins_to_constant_network(conn, vcc_track_pkey, gnd_track_pkey):
+    """
+    Connect all GND_WIRE* and VCC_WIRE* nodes to VCC or GND track.
+    """
+
+    cur = conn.cursor()
+
+    write_cur = conn.cursor()
+    write_cur.execute("""BEGIN EXCLUSIVE TRANSACTION;""")
+
+    # VCC
+    cur.execute(
+        """
+SELECT pkey FROM wire_in_tile
+WHERE name LIKE "VCC_WIRE%"
+"""
+    )
+
+    for (wire_in_tile_pkey, ) in cur:
+        write_cur.execute(
+            """
+UPDATE node SET track_pkey = ? WHERE pkey IN (
+    SELECT node_pkey FROM wire WHERE wire_in_tile_pkey = ?
+)
+            """, (
+                vcc_track_pkey,
+                wire_in_tile_pkey,
+            )
+        )
+
+    # GND
+    cur.execute(
+        """
+SELECT pkey FROM wire_in_tile
+WHERE name LIKE "GND_WIRE%"
+"""
+    )
+
+    for (wire_in_tile_pkey, ) in cur:
+        write_cur.execute(
+            """
+UPDATE node SET track_pkey = ? WHERE pkey IN (
+    SELECT node_pkey FROM wire WHERE wire_in_tile_pkey = ?
+)
+            """, (
+                gnd_track_pkey,
+                wire_in_tile_pkey,
+            )
+        )
+
+    write_cur.execute("""COMMIT TRANSACTION""")
+
+
+def identify_const_nodes(conn):
+    """
+    Creates a new table for const nodes, identifies them and inserts their
+    pkeys along with const signal name to the table.
```

Comment:
This docstring is insufficient.  In generally all of the docstrings in this file need to be expanded.

Reply:
I've updated docstrings. Please point me out where else you find them to be insufficient.